ts/src/coinmetro.ts

```diff
@@ -0,0 +1,1922 @@
+        //             "notabeneDecimals": 8,
+        //             "canMarket": true,
+        //             "maxSwap": 10000,
+        //             "digits": 6,
+        //             "multiplier": 1000000,
+        //             "bookDigits": 8,
+        //             "bookMultiplier": 100000000,
+        //             "sentimentData": {
+        //                 "sentiment": 51.59555555555555,
+        //                 "interest": 1.127511216044664
+        //             },
+        //             "minQty": 0.0001
+        //         },
+        //         {
+        //             "symbol": "EUR",
+        //             "name": "Euro",
+        //             "color": "#1246FF",
+        //             "type": "fiat",
+        //             "canDeposit": true,
+        //             "canWithdraw": true,
+        //             "canTrade": true,
+        //             "canMarket": true,
+        //             "maxSwap": 10000,
+        //             "digits": 2,
+        //             "multiplier": 100,
+        //             "bookDigits": 3,
+        //             "bookMultiplier": 1000,
+        //             "minQty": 5
+        //         }
+        //         ...
+        //     ]
+        //
+        const result = {};
+        for (let i = 0; i < response.length; i++) {
+            const currency = response[i];
+            const id = this.safeString (currency, 'symbol');
+            const code = this.safeCurrencyCode (id);
+            const withdraw = this.safeValue (currency, 'canWithdraw');
+            const deposit = this.safeValue (currency, 'canDeposit');
+            const canTrade = this.safeValue (currency, 'canTrade');
+            // todo: check active limits and precision (what are notabeneDecimals, digits and bookDigits?)
+            const active = canTrade ? withdraw : true;
+            const precision = this.safeInteger (currency, 'digits');
+            const minAmount = this.safeNumber (currency, 'minQty');
+            result[code] = this.safeCurrencyStructure ({
+                'id': id,
+                'code': code,
+                'name': code,
+                'info': currency,
+                'active': active,
+                'deposit': deposit,
+                'withdraw': withdraw,
+                'fee': undefined,
+                'precision': precision,
+                'limits': {
+                    'amount': { 'min': minAmount, 'max': undefined },
+                    'withdraw': { 'min': undefined, 'max': undefined },
+                },
+                'networks': {},
+            });
+        }
+        // todo: check
+        if (this.safeValue (this.options, 'currenciesByIdForParseMarket') === undefined) {
+            const currenciesById = this.indexBy (result, 'id');
+            this.options['currenciesByIdForParseMarket'] = currenciesById;
+            this.options['currencyIdsListForParseMarket'] = Object.keys (currenciesById);
```

Comment:
Why do we need this auxiliary structure? 

`this.currencies` stores the currencies using the unified currency code right?, for example: 'BTC: '{}` 

Let's say that inside `parseMarketPrecisionAndLimits` you need to get the currency using the exchange-specific currency ID, can't you just do `const code = this.safeCurrencyCode(currencyId)` and then `const currencyInfo = this.currencies[code]` ?

Reply:
Same thing for the `currencyIdsListForParseMarket` structure

Reply:
`this.currencies` has value only after the call of `loadMarkets`. If we call `fetchMarkets` before it, markets could not be parsed cuz `this.currencies` is yet undefined. I could write something like this in `fetchMarkets`:
```js
if (this.currecies === undefined) {
    return await this.loadMarkets ();
}
```
Also `this.options['currencyIdsListForParseMarket']` is suitable for `parseMarket` for not to use `Object.keys (this.currencies_by_id)` in every iteration of the loop of `parseMarkets`

Reply:
`fetchMarkets` and `loadMarkets` have return values, so we can't use `return await this.loadMarkets ()` inside of `fetchMarkets`. Since I can't find the way how to get info about currencies without these auxiliary structures.

Reply:
ok I think we can leave it as is for now